### ISSUE TITLE
NewlineAfterOpenBrace being applied to PSR2 incorrectly.

### ISSUE THREAD

**Issue by brianlmoon** [squizlabs/PHP_CodeSniffer#520]
Hi,

I have code like:
```php
        if ($log_errors || $display_errors) {

            $error = array(
                "timestamp" => $this->getTimestamp(),
                "message" => $errstr,
                "errno" => $errno,
                "source_file" => $errfile,
                "source_line" => $errline
            );
            // more and more code
        }
```
Of cours phpcs complains with:
```
Expected 1 newline after opening brace; 2 found
(Squiz.ControlStructures.ControlSignature.NewlineAfterOpenBrace)
```
However, I can find nothing in the PSR1 or [PSR2](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#5-control-structures) spec that says this is invalid.

**Comment by gsherwood** [comment]
The example code there shows the IF body on the line after the declaration. In fact, the PSR2 standard *only* uses example code to show how IF statements should specifically be defined.

PSR2 is also fairly ambiguous in many places because the authors didn't consider ever possible piece of code. If you don't agree with me, one thing you can do is go and ask about this on the mailing list and see if the original authors agree with you. If they make a call on it, I will change PHPCS to allow any number of empty lines at the top of control structures.

The other thing you can do is just create your own coding standard using a ruleset.xml file where you mute this particular error message. To do that, create a file called ruleset.xml with the following contents:
```xml
<?xml version="1.0"?>
<ruleset name="MyStandard">
 <description>My custom coding standard.</description>
 <rule ref="PSR2">
  <exclude name="Squiz.ControlStructures.ControlSignature.NewlineAfterOpenBrace"/>
 </rule>
</ruleset>
```
Then use it to checkout your code: `phpcs --standard=/path/to/ruleset.xml /path/to/code`

There is more info about what you can do in ruleset files here: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml

**Comment by brianlmoon** [comment]
Thanks for the reply Greg.

I have posted to the list.

**Comment by harikt** [comment]
Hey @brianlmoon , 

As there were more comments regarding PSR-1 and PSR-2, there is also a dedicated group  https://groups.google.com/forum/#!forum/php-fig-cs

**Comment by mikehall314** [comment]
Would section 2.3 cover this case? "Blank lines MAY be added to improve readability and to indicate related blocks of code."?

**Comment by brianlmoon** [comment]
@mikehall314 Thank you. I knew I had read that somewhere in the spec and could not find it. That, along with comments on the mailing list, would seem to indicate at worst, this is not addressed at all by the spec and should be removed from the phpcs build for psr2.

**Comment by gsherwood** [comment]
> Would section 2.3 cover this case? "Blank lines MAY be added to improve readability and to indicate related blocks of code."?

I'm pretty sure that is talking about code blocks and not about defined control structures that have sample code provided to show their layout. Otherwise, it would directly conflict with all the other parts of the standard that say things like `the closing brace for the class MUST go on the next line after the body`. So I can add a few lines before the closing brace because another part of the standard says I can if I think it improves readability? If not, then you're placing some parts of the standard above others. Specifically, you're saying that the code examples are not actually that important despite saying `Note the placement of parentheses, spaces, and braces`.

If you want to, you can interpret PSR-2 as "add blank lines and comments anywhere you want" because it kind of says these things in parts. And the use of code examples instead of writing out rules using words leaves everything ambiguous because that says "if your code doesn't look like this, it isn't PSR-2 compliant".

So there is never going to be a winner in an argument over PSR-2 because it's not a properly written standard. It's not that they didn't try to write one; it's just damn hard when everyone decides to use it and PHP is so flexible.

The original interpretation of this rule that I implemented, and provided for checking by PHP-FIG before release, is the one I am going to continue to use within PHPCS. I'm not going to relax the PSR2 standard in PHPCS even further by removing all checks that have anything to do with line spacing, or to disregard the formatting provided in the code examples.

I encourage people to write their own coding standard using a ruleset.xml file and exclude bits you don't like (and include other checks you do). Don't worry so much about an official PSR-2 tick of approval. Even the member projects clearly don't care about that.
